### ISSUE TITLE
Recurrence missing 'until' option

### ISSUE THREAD

**Issue by DirkReiners** [kuzmoyev/Google-Calendar-Simple-API#9]
Recurrence handling is missing one of the most important options: giving an end date for the repeating event (UNTIL).

The following patch works for me, but can probably take some polishing:

```
$ diff -c3 recurrence.py recurrence.py.orig
*** recurrence.py       2020-05-24 00:27:54.917591200 -0400
--- recurrence.py.orig  2020-05-23 23:50:52.821505600 -0400
***************
*** 122,128 ****
              freq=DAILY,
              interval=None,
              count=None,
-             until=None,
              by_second=None,
              by_minute=None,
              by_hour=None,
--- 122,127 ----
***************
*** 143,150 ****
                  positive integer representing how often the recurrence rule repeats
          :param count:
                  number of occurrences at which to range-bound the recurrence
-         :param until:
-                 end date of recurrence
          :param by_second:
                  second or list of seconds within a minute. Valid values are 0 to 60
          :param by_minute:
--- 142,147 ----
***************
*** 182,188 ****

          .. _`RRULE format`: https://tools.ietf.org/html/rfc5545#section-3.8.5
          """
!         return 'RRULE:' + Recurrence._rule(freq, interval, count, until, by_second, by_minute, by_hour, by_week_day,
                                             by_month_day, by_year_day, by_week, by_month, by_set_pos, week_start)

      @staticmethod
--- 179,185 ----

          .. _`RRULE format`: https://tools.ietf.org/html/rfc5545#section-3.8.5
          """
!         return 'RRULE:' + Recurrence._rule(freq, interval, count, by_second, by_minute, by_hour, by_week_day,
                                             by_month_day, by_year_day, by_week, by_month, by_set_pos, week_start)

      @staticmethod
***************
*** 190,196 ****
              freq=DAILY,
              interval=None,
              count=None,
-             until=None,
              by_second=None,
              by_minute=None,
              by_hour=None,
--- 187,192 ----
***************
*** 211,218 ****
                  positive integer representing how often the recurrence rule repeats
          :param count:
                  number of occurrences at which to range-bound the recurrence
-         :param until:
-                 end date of recurrence
          :param by_second:
                  second or list of seconds within a minute. Valid values are 0 to 60
          :param by_minute:
--- 207,212 ----
***************
*** 250,256 ****

          .. _`RRULE format`: https://tools.ietf.org/html/rfc5545#section-3.8.5
          """
!         return 'EXRULE:' + Recurrence._rule(freq, interval, count, until, by_second, by_minute, by_hour, by_week_day,
                                              by_month_day, by_year_day, by_week, by_month, by_set_pos, week_start)

      @staticmethod
--- 244,250 ----

          .. _`RRULE format`: https://tools.ietf.org/html/rfc5545#section-3.8.5
          """
!         return 'EXRULE:' + Recurrence._rule(freq, interval, count, by_second, by_minute, by_hour, by_week_day,
                                              by_month_day, by_year_day, by_week, by_month, by_set_pos, week_start)

      @staticmethod
***************
*** 399,405 ****
              freq=DAILY,
              interval=None,
              count=None,
-             until=None,
              by_second=None,  # BYSECOND
              by_minute=None,  # BYMINUTE
              by_hour=None,  # BYHOUR
--- 393,398 ----
***************
*** 420,427 ****
                  positive integer representing how often the recurrence rule repeats
          :param count:
                  number of occurrences at which to range-bound the recurrence
-         :param until:
-                 end date of recurrence
          :param by_second:
                  second or list of seconds within a minute. Valid values are 0 to 60
          :param by_minute:
--- 413,418 ----
***************
*** 490,502 ****
              raise ValueError('"count" parameter must be a positive int. '
                               '{} was provided'.format(count))

-         if until:
-             if not isinstance(until, (date, datetime)):
-                 msg = 'The until object must be a date or datetime, not {!r}.'.format(end.__class__.__name__)
-                 raise TypeError(msg)
-             else:
-                 until = until.strftime("%Y%m%dT%H%M%SZ")
-
          by_second = assure_iterable(by_second)
          check_all_type_and_range(by_second, int, (0, 60), "by_second")

--- 481,486 ----
***************
*** 537,543 ****
          rule_properties = (
              ('INTERVAL', interval),
              ('COUNT', count),
-             ('UNTIL', until),
              ('BYSECOND', to_string(by_second)),
              ('BYMINUTE', to_string(by_minute)),
              ('BYHOUR', to_string(by_hour)),
--- 521,526 ----
```

Hope it helps!

**Comment by kuzmoyev** [comment]
The initial thought was to avoid having `UNTIL` in recurrence and have it only in `end_date` of `Event` for simplification. I believe having `UNTIL` specified is the same as having `end` specified in the body of the event.

But the fact that you feel like this parameter is missing means that it might be a flaw from UX point:)

Thank you for the submission of the issue.
I'd be grateful if you could submit your changes as a pull request so I could take a proper look at it and merge it into the `dev` branch.

**Comment by DirkReiners** [comment]
Hm, I don't think end_date covers that. Let's say you have an event from 23:30 to 00:30 every Tuesday until June 3rd the start and end dates for the initial event will be different, and can't be June 3rd, or you'll have an event from now until June 3rd.

I'll look into making a pull request, haven't done the before.

**Comment by kuzmoyev** [comment]
Yes, I think you're right. We will need to add it. If you have time to create a pull request please do it to `dev` branch. If you don't, I'll add it later when I have time.

Thank you anyway

**Comment by DirkReiners** [comment]
Hi Yevhen,

just submitted the pull request.

Hope it helps!

  Dirk

**Comment by kuzmoyev** [comment]
Closed by [PR #10](https://github.com/kuzmoyev/Google-Calendar-Simple-API/pull/10)
